src/d3a/models/strategy/external_strategies/storage.py

```diff
@@ -246,14 +253,17 @@ def _bid(self, payload):
             return
         try:
             arguments = json.loads(payload["data"])
-            assert set(arguments.keys()) == {'price', 'energy', 'transaction_id'}
+            assert all(arg in {'price', 'energy', 'replace_existing'} for arg in arguments.keys())
```

Comment:
I know that this is starting the discussion of ALLOWED and NEEDED parameters again, but `price` and `energy` are actually mandatory parameters, otherwise no bid can be placed.

Reply:
I could add something like this inside the `_bid` method (rather than as class variables):
```python
required_args = {'price', 'energy'}
allowed_args = required_args.union({'replace_existing', 'transaction_id'})

# Check that all required arguments have been provided
assert all(arg in arguments.keys() for arg in required_args)
# Check that every provided argument is allowed
assert all(arg in allowed_args for arg in arguments.keys())
```
What do you think?

In any case, I think this logic should go in a separate validator class (or similar) in the future. Maybe we could open a ticket for it.